### PR TITLE
Enable the Opus music encoder

### PR DESCRIPTION
**Pull request by SuperNascher**
This commit adds the ability to use the AUDIO encoder
instead of the VOIP encoder, that can be used for
bots or broadcaster, who wants to stream in a higher quality.

To enable this feature add this to the Mumble config:
[codec]
opus/encoder/music=true

Fixes mumble-voip/mumble#1630

### CODE REVIEW

File: src/mumble/Settings.cpp [hunk 1]
```diff
@@ -878,6 +881,7 @@ void Settings::save() {
 	SAVELOAD(vsVAD, "audio/vadsource");
 	SAVELOAD(fVADmin, "audio/vadmin");
 	SAVELOAD(fVADmax, "audio/vadmax");
+        SAVELOAD(bUseOpusMusicEncoding, "audio/useopusmusicencoding");
```
Comment (bendem) [inline src/mumble/Settings.cpp]: spaces

File: src/mumble/AudioInput.cpp [hunk 2]
```diff
@@ -85,7 +85,14 @@ AudioInput::AudioInput() : opusBuffer(g.s.iFramesPerPacket * (SAMPLE_RATE / 100)
 	iFrameSize = SAMPLE_RATE / 100;
 
 #ifdef USE_OPUS
-	opusState = opus_encoder_create(SAMPLE_RATE, 1, OPUS_APPLICATION_VOIP, NULL);
+	if (!g.s.bUseOpusMusicEncoding) {
+		opusState = opus_encoder_create(SAMPLE_RATE, 1, OPUS_APPLICATION_VOIP, NULL);
+		qWarning("Opus encoder set for VOIP");
```
Comment (mkrautz) [inline src/mumble/AudioInput.cpp]: please prefix the log entries with "AudioInput: "

File: src/mumble/AudioInput.cpp [hunk 3]
```diff
@@ -85,7 +85,14 @@ AudioInput::AudioInput() : opusBuffer(g.s.iFramesPerPacket * (SAMPLE_RATE / 100)
...
+	} else {
+		opusState = opus_encoder_create(SAMPLE_RATE, 1, OPUS_APPLICATION_AUDIO, NULL);
+		qWarning("Opus encoder set for Music");
```
Comment (mkrautz) [inline src/mumble/AudioInput.cpp]: Ditto.

File: src/mumble/Settings.cpp [hunk 4]
```diff
@@ -567,6 +569,7 @@ void Settings::load(QSettings* settings_ptr) {
 	LOADENUM(vsVAD, "audio/vadsource");
 	SAVELOAD(fVADmin, "audio/vadmin");
 	SAVELOAD(fVADmax, "audio/vadmax");
+	SAVELOAD(bUseOpusMusicEncoding, "audio/useopusmusicencoding");
```
Comment (mkrautz) [inline src/mumble/Settings.cpp]: I'd still prefer "codec/opus/encoder/music" as I mentioned on IRC. Thoughts?

**Comment by mkrautz** [comment]
LGTM beside my comments.

**Comment by SuperNascher** [comment]
Ok

File: src/mumble/Settings.cpp [hunk 5]
```diff
@@ -901,6 +905,8 @@ void Settings::save() {
 	SAVELOAD(bWhisperFriends, "audio/whisperfriends");
 	SAVELOAD(bTransmitPosition, "audio/postransmit");
 
+        SAVELOAD(bUseOpusMusicEncoding, "codec/opus/encoder/music");
```
Comment (mkrautz) [inline src/mumble/Settings.cpp]: Spaces (should be tabs)

File: src/mumble/Settings.cpp [hunk 6]
```diff
@@ -590,6 +592,8 @@ void Settings::load(QSettings* settings_ptr) {
...
```
Comment (mkrautz) [inline src/mumble/Settings.cpp]: Spaces (should be tabs)

**Comment by TFMvdBroek** [comment]
Music quality does not seem to improve when i add "opus\encoder\music=true" under [codec] in my bots mumble.ini. Running version 1.2.17. 
Any clue?

**Comment by SuperNascher** [comment]
@TFMvdBroek It will be available, when Mumble 1.3 is released. If you want to use the encoder now, you must use the Snapshot or the Git version of Mumble.

**Comment by TFMvdBroek** [comment]
Running latest snapshot also does not seem to improve quality. Any way to check if the setting is actually enabled?

**Comment by Geraner** [comment]
The latest snapshot version 1533 (static linux server) does not support this change yet. New snapshot 1634 needs to be created.

**Comment by Natenom** [comment]
@Geraner This is a client only change. Server does not need to support it.

**Comment by TFMvdBroek** [comment]
I am running the latest Windows snapshot, which should work... But no difference in music quality tho

**Comment by TFMvdBroek** [comment]
I am running the latest Windows snapshot, which should work... But no difference in music quality tho

**Comment by Oatwalker** [comment]
I've been trying to get studder free music playing in mumble for ages.  It's been really frustrating.  I'll have to try the newest snapshot and see if it helps.

**Comment by bbigelow0621** [comment]
Hi, I am a n00b when it comes to coding. Can someone please explain in layman terms how to add this feature to Mumble. Thank youuuuu!

**Comment by uncivilgentleman** [comment]
> k

I can get pretty good music quality by simple increasing the max bit-rate on the server.  Then making sure the client streaming the music also has a high bit-rate and turning off Noise Suppression, Max Amplification, and RNNoise (just the client where the music is coming from).

Keep in mind it'll still be single channel audio.  Supposedly stereo sound is coming with Mumble 1.4.0.

Let me know if you have any more questions and I'll try to answer them.
